### PR TITLE
SWIK 655 1068 837 1114 1188 1193 1253 1265 1266 1267 1270 1300 1302 1301 1268 1272 1254 1286 655

### PR DESCRIPTION
**Pull request by kadevgraaf**
This is a large PR for an overhaul of the slideContentEditor interaction with a right-click context menu, on-click drag and resize, and accesskeys

Addresses several SWIK issue:
- https://slidewiki.atlassian.net/browse/SWIK-655 - Can't alter properties of an image on an imported slide
- https://slidewiki.atlassian.net/browse/SWIK-1068 - SVG control box in slide edit flashes/keeps losing focus
- https://slidewiki.atlassian.net/browse/SWIK-837 - Create drag and resize boxes for SVG graphics
- https://slidewiki.atlassian.net/browse/SWIK-1114 - Remove selection of input box controls from draggable.js
- https://slidewiki.atlassian.net/browse/SWIK-1188 - Add unique id attributes to HTML elements of a slide in slide editor
- https://slidewiki.atlassian.net/browse/SWIK-1193 - remove CKeditor from title in slide content - optimize/perfomance
- https://slidewiki.atlassian.net/browse/SWIK-1253 - Send to back button not always working
- https://slidewiki.atlassian.net/browse/SWIK-1265 - Onmouseover input box - higlight
- https://slidewiki.atlassian.net/browse/SWIK-1266 - Mimick input box resize/drag/select interaction from well-established presentation tools
- https://slidewiki.atlassian.net/browse/SWIK-1267 - Right-click context-menu for move-to-front and move-to-back of input boxes 
- https://slidewiki.atlassian.net/browse/SWIK-1270 - move slide area of pptx2html div to right - center (top/bottom) - similar to other presentation tools
- https://slidewiki.atlassian.net/browse/SWIK-1300 - Use single click and Double click to switch between dragging and editing content in input box + Remove drag button.
- https://slidewiki.atlassian.net/browse/SWIK-1302 - Create thin line around input boxes - similar to other editing tools 
- https://slidewiki.atlassian.net/browse/SWIK-1301 - Resize using borders of input box (instead of bottom-right resize button)
- https://slidewiki.atlassian.net/browse/SWIK-1268 - Use library to mimick image drag/resize/crop interaction of well-established presentation/image tools
- https://slidewiki.atlassian.net/browse/SWIK-1272 - Simple and advanced edit mode - two CKeditor configs?
- https://slidewiki.atlassian.net/browse/SWIK-1254 - Text cursor is located to the left of text in input boxes
- https://slidewiki.atlassian.net/browse/SWIK-1286 - Remove right-click context menu in CKeditor - and/or use own buttons/menu
- https://slidewiki.atlassian.net/browse/SWIK-655 - Can't alter properties of an image on an imported slide

To check:
- Is this good for usability?
- Is this good for accessibility? (tabbing between elements has (some?/good? support). important accesskeys are denoted in right-click context menu (which can be opened with "ctrl+alt" or "alt+ctrl"). I am not sure what to do for screen readers and aria-label. 
- (related to above:) Does it work on mobile with long press?

To do / to fix: 
- Nested Div elements, see slide http://localhost:3000/deck/1643-1/slide/12644-1/12644-1:5/edit (no edit of text in nested div element + no resize + onclick to edit gives image dialog)
- Caret (key cursor) position in firefox keeps reseting on mouseclick/drag - can use keyboard cursor
- CKeditor size and style dropdown does not work
- SVG resize (can drag)
- firefox onmouseover border is not accurate

### CODE REVIEW

**Comment by abijames** [comment]
Lots of goof improvements. Seems more reliable. I still sometimes lots the input box functionality but I think this was focus issues. As this refers to so many tickets, I'm putting the comments for all of them here...
Moving simple content around is good but need keyboard method for moving boxes, Check Google slides shortcuts - https://support.google.com/docs/answer/1696717?hl=en
* Ctrl page up/down for reorder is good instead of Ctrl + arrow. Do not use Ctrl + arrow key as conflicts with screen readers
* Suggest that to move a box just use the arrows once in focus (as with google slides)

Difficult to know when you are in edit mode vs move mode for a box. Can't use only colour to distinguish this .
* Need a keyboard or button function for exiting edit (e.g. ctrl-enter)
* Boundaries only appear on hover-over…. Should be on when the box has focus. This would then allow for implementing the arrow keys to move an input box.
* Should be a contrasting colour to the background e.g Google slides uses #0096FD with a lighter border which seems to work well as unlikely to be a popular background choice!

The button to select the advanced toolbar needs to be more obvious. But if in RH panel (design tab) it will help as it will be more closely located to the toolbar.
 
Ctrl+enter when in edit mode, the outline of input box disappears. I think this is related to the outline of the box only appearing when the mouse is hovering over the box. So again may be addressed by moving to a box being selected when it has focus.
 
When I click on an input box that extends below the scroll line the whole page scrolls down - distracting and loose an indication of my location.
 
I like that if I double click on a slide it takes me into the edit mode - is that deliberate?
 
When I delete an element (an input box) I want to be able to undo (ctrl-z)!
 
Speaker's notes - only needs simple toolbar 
 
Touchscreen - this needs further testing as it was on an iPad duplicating a PC screen but a good indication that there are some issues:
* Context menu did not appear on long press
* When select text then context menu appeared as a popup but would not activate with touch
 
Still confusing when content in a box appears outside the margins of the input box e.g. http://192.168.56.101:3000/deck/1490/slide/10832-5/10832-5:1/edit

Did not test uploading images as I am aware this will change. Will now attempt to test with a screen reader.

**Comment by abijames** [comment]
Screen reader compatibility - this is promising (getting announcements about editable text boxes) but it won't work correctly until editing input boxes and focus are aligned as well as the keyboard shortcuts

File: assets/ckeditor_config_basic.js [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+	// Remove some buttons provided by the standard plugins, which are
+	// not needed in the Standard(s) toolbar.
+
+    //klaas adapt: extra right-click menu options
+    //config.extraPlugins = 'dialogadvtab';
+
+	//config.removeButtons = 'Underline,Subscript,Superscript';
+
+	// Set the most common block elements.
+    config.format_tags = 'p;h1;h2;h3;pre';
```
Comment (bwulff) [inline assets/ckeditor_config_basic.js]: This has already been changed by Umar's commit https://github.com/slidewiki/slidewiki-platform/commit/b5d875d6d40870cf5ea6fba121ee3bc8b9f07293

**Comment by abijames** [comment]
@kadevgraaf are you ready for me to review your latest commits?

**Comment by kadevgraaf** [comment]
Hi @abijames and @bwulff !
Many thanks for your feedback!
@abijames  - 

> Moving simple content around is good but need keyboard method for moving boxes, Check Google slides shortcuts

I tried to implement this, see the keymap on https://github.com/slidewiki/slidewiki-platform/pull/403/files#diff-25795d946073d72c263dd612ced7f802R1052
- Should we make the keymaps explicit somewhere? Or would it suffice to do this in a video guide?
- should I try to adopt (more of) the google slides shortcuts?

> Ctrl+enter when in edit mode, the outline of input box disappears. I think this is related to the outline of the box only appearing when the mouse is hovering over the box. So again may be addressed by moving to a box being selected when it has focus.
I think I fixed this. Can you check again?

> I like that if I double click on a slide it takes me into the edit mode - is that deliberate?
Yes, very much so :)

> Speaker's notes - only needs simple toolbar
Done

> Did not test uploading images as I am aware this will change. Will now attempt to test with a screen reader.
I did improve the selection of images in keymode (a span is added so you can navigate to an image to see the image details in CKeditor (with alt + F10)

TODO - 
> When I delete an element (an input box) I want to be able to undo (ctrl-z)!

> When I click on an input box that extends below the scroll line the whole page scrolls down - distracting and loose an indication of my location.

>Touchscreen - this needs further testing as it was on an iPad duplicating a PC screen but a good indication that there are some issues:
>    Context menu did not appear on long press
>    When select text then context menu appeared as a popup but would not activate with touch

@bwulff 
I found a way to overlay the right-click context menu with the CKeditor right-click context menu so this'll revert the change on hindsight :)

**Comment by abijames** [comment]
> I tried to implement this, see the keymap on https://github.com/slidewiki/slidewiki-platform/pull/403/files#diff-25795d946073d72c263dd612ced7f802R1052
>Should we make the keymaps explicit somewhere? Or would it suffice to do this in a video guide?
should I try to adopt (more of) the google slides shortcuts?
Shortcut keys for moving the input box boxes need to be different as the conflict with browser zoom tools. Suggest ctrl+alt+ +/-.
We will list keyboard shortcuts in the help guide.


Re focus of the input boxes, this works correctly but propose that on slides that use a template the slide area (i.e. for document layout) is removed from the tab order. This means that users will tab directly to the first input box.

ctrl+enter is working correctly to allow users to edit a input box Exiting editing an input box needs a shortcut key. Propose ESC

RE touchscreen.... now if I double tap an input box I see the context menu but then it is quickly replaced with ckeditor context menu. I think this needs to be tested on different devices once merged with master to check behaviour on different OS.

**Pull request merged**